src/analyses/ai_history.h

```diff
@@ -0,0 +1,227 @@
+/*******************************************************************\
+
+Module: Abstract Interpretation
+
+Author: Martin Brain, martin.brain@cs.ox.ac.uk
+
+\*******************************************************************/
+
+/// \file
+/// Abstract Interpretation history
+///
+/// This is an abstraction of the history of the program counter
+/// / the history of the execution of the program.
+/// It is used to track and control how the abstract interpreter
+/// explores the program and for context sensitive analyses it is
+/// also used as the key for the state map.
+///
+/// The abstract interpreter works with instances of ai_history_baset
+/// (and history_ptrt's when dealing with storage) but this is just
+/// an abstract interface much like ai_domain_baset, actual implementations
+/// derive from it.
+
+#ifndef CPROVER_ANALYSES_AI_HISTORY_H
+#define CPROVER_ANALYSES_AI_HISTORY_H
+
+#include <memory>
+
+#include <goto-programs/goto_model.h>
+
+#include <util/json.h>
+#include <util/xml.h>
+
+/// We make heavy use of reference counted pointers to history objects
+class ai_history_baset;
+
+class history_ptrt
+{
+public:
+  typedef std::shared_ptr<const ai_history_baset> ptrt;
+
+  history_ptrt(ptrt _p) : p(_p)
+  {
+  }
+
+  history_ptrt(const ai_history_baset *h) : p(h)
+  {
+  }
+
+  // Order and compare on the contents not the pointers
+  bool operator<(const history_ptrt &op) const;
+  bool operator==(const history_ptrt &op) const;
+
+  const ai_history_baset &operator*(void) const
+  {
+    return *p;
+  }
+
+private:
+  ptrt p;
+};
+
+/// This is the base interface; derive from this.
+class ai_history_baset
+{
+public:
+  typedef goto_programt::const_targett locationt;
+  typedef irep_idt function_namet;
+
+  ai_history_baset(const ai_history_baset &)
+  {
+  }
+
+  /// Create a new history starting from a given location
+  /// This is used to start the analysis from scratch
+  /// PRECONDITION(l.is_dereferenceable());
+  explicit ai_history_baset(locationt)
+  {
+  }
+
+  enum class step_statust
```

Comment:
A bit of doxygen explaining the below would be helpful.

Reply:
Explained in step below.  Is there a way to view the CI generated doxygen to check it has come out OK?

Reply:
You can just run `doxygen` in `src` locally and then open `doc/html/index.html`

Reply:
Done and now it is correctly linked to the right functions.